relay-server/src/processing/client_reports/process.rs

```diff
@@ -0,0 +1,307 @@
+use std::collections::BTreeMap;
+use std::error::Error;
+
+use chrono::{Duration as SignedDuration, Utc};
+use relay_common::time::UnixTimestamp;
+use relay_config::Config;
+use relay_event_normalization::ClockDriftProcessor;
+use relay_event_schema::protocol::ClientReport;
+use relay_filter::FilterStatKey;
+use relay_quotas::ReasonCode;
+use relay_sampling::evaluation::MatchedRuleIds;
+use relay_system::Addr;
+
+use crate::constants::DEFAULT_EVENT_RETENTION;
+use crate::managed::Managed;
+use crate::processing::client_reports::SerializedClientReport;
+use crate::services::outcome::{Outcome, RuleCategories, TrackOutcome};
+use crate::services::processor::MINIMUM_CLOCK_DRIFT;
+use crate::services::projects::project::ProjectInfo;
+
+/// Fields of client reports that map to specific [`Outcome`]s without content.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ClientReportField {
+    /// The event has been filtered by an inbound data filter.
+    Filtered,
+
+    /// The event has been filtered by a sampling rule.
+    FilteredSampling,
+
+    /// The event has been rate limited.
+    RateLimited,
+
+    /// The event has already been discarded on the client side.
+    ClientDiscard,
+}
+
+/// Validates and extracts client reports.
+///
+/// At the moment client reports are primarily used to transfer outcomes from
+/// client SDKs.  The outcomes are removed here and sent directly to the outcomes
+/// system.
+pub fn process_client_reports(
```

Comment:
This logic is taken almost 1-1 from the old processing so if we think this is better broken up or rewritten to make it more elegant can still do that.

Reply:
Ok, rewrote the logic now to split it up ruffly into the individual parts that you mentioned. Not sure if the code is too cute now though.